### DISCUSSION TITLE
wger docker with Traefik

### DISCUSSION THREAD

**Discussion by hubortje** [wger-project/docker#96] [Q&A]
Hello,

does anybody got a working configuration using wger docker with Traefik?
I'm heavily confused by the need for the nginx-config, and I'm struggling to set it up with Traefik.

Best regards

**Comment by rolandgeider** [comment]
hi! The reason for nginx is that django doesn't serve static files during production. If you can configure traefik as a reverse proxy and make it serve the files, everything should work fine (but you will have to make sure they are either in the same network so that things like `web:8000` work or open the appropriate ports)

**Comment by hubortje** [comment]
Moin @rolandgeider,

thanks for your quick response.
I've taken the time at work to get to understand the problem better, and found a similar issue in the tracker.

The static files are now the issue, like you've mentioned.
Django does not really like me accessing the web-server, as I'm faced with a page without statics, and cannot log in. `DEBUG=True` only offers a 

> CSRF-Verifizierung fehlgeschlagen. Anfrage abgebrochen.

or probably translated to

> CSRF verification failed. Request cancelled.

As mentioned in this reply https://github.com/wger-project/docker/issues/32#issuecomment-1985198762 it seems like the issue is somewhere in this line of communication
> So Users -> Nginx Port -> Django Port and Users -> Nginx Port -> Filesystem needs to work to serve the pages.

I'm not sure when I will find time to continue to wrap my head around this issue, but I'll try to document it here, whenever.

Also, did I understand correctly, that celery is optional, if I don't need regularly synchronization of ingredients, exercises, and videos?

**Comment by hubortje** [comment]
Some errors I've currently got going:

From the Postgres log:
`FATAL:  role "wger" does not exist`

from the web-server log:
`WARNING 2024-08-11 15:08:41,995 log Forbidden (Origin checking failed - https://my.domain.com does not match any trusted origins.): /de/user/login`

Furthermore, `docker exec wger-web env` prints

```
PATH=/home/wger/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=f6ddf2d29da7
DEBIAN_FRONTEND=noninteractive
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_ALL=en_US.UTF-8
PYTHONDONTWRITEBYTECODE=1
PYTHONUNBUFFERED=1
HOME=/home/wger
```
while I'd expect it to print at least one of the variables defined in
```
env_file:
     - ./wger.env
```

**Comment by hubortje** [comment]
Ok, wild ride.

So, I've started by investigating the `FATAL:  role "wger" does not exist` error from the Postgres log.
Found out, that the user declared in the docker-compose.yml [here](https://github.com/wger-project/docker/blob/ec159f0af9a49c8472fd58c6eddd81f0b15ee29e/docker-compose.yml#L52C5-L55C13), has to be the same as declared in the .env-file [here](https://github.com/wger-project/docker/blob/ec159f0af9a49c8472fd58c6eddd81f0b15ee29e/config/prod.env#L89).

I'll need to investigate again, if I can use a different username and password as `wger`, but it seemed (didn't track it, will do and change this paragraph) to not work, if I change the user from `wger` to e.g. `wgeruser`.
Why not work with a $-placeholder in the docker-compose.yml that and let the user, password, and db-name be declared in the .env-file?

So, currently my db is running as `wger`, with the user and password `wger`.

Still had trouble with the statics. They weren't moved to my `${DATA}/wger/data/static` path, but were present inside the container in `/home/wger/static/`.
Thanks to [this reply/issue](https://github.com/wger-project/docker/issues/92#issuecomment-2243192333), I've figured out to copy the files from the container to my disk by using `docker cp wger-container-name:/home/wger/static/. /path/to/data/disk/wger/data/static`.

Note, that the `/.` at the end of `wger-container-name:/home/wger/static` is needed by `docker cp` to copy all files from the location. No `/.` at the end will suggest to copy only a file from `home/wger/static` with no name.

Now the statics are visible when visiting the webpage.

EDIT: This is not a long-term solution. After a restart of the container, the statics are not showing again, ~~despite being in the proper location on the disc~~, as they have to be re-copied.

**Reply by hubortje** [reply]
> I'll need to investigate again, if I can use a different username and password as `wger`, but it seemed (didn't track it, will do and change this paragraph) to not work, if I change the user from `wger` to e.g. `wgeruser`. Why not work with a $-placeholder in the docker-compose.yml that and let the user, password, and db-name be declared in the .env-file?
> 
> So, currently my db is running as `wger`, with the user and password `wger`.

Ok, so the user `wger` is created on startup of the db. Why not add it as a variable, which is then taken by the .env?
So that in the [.env-file](https://github.com/wger-project/docker/blob/ec159f0af9a49c8472fd58c6eddd81f0b15ee29e/config/prod.env#L89), one could declare the user, password, and db-name, and this information is used during container start up?
